src/offset/library.cairo

```diff
@@ -106,68 +115,103 @@ namespace CarbonableOffseter {
-        );
+        // [Check] Contract token id is not null
+        let (token_id) = _get_token_id();
+        let zero = Uint256(low=0, high=0);
+        let (equal) = uint256_eq(token_id, zero);
+        if (equal == TRUE) {
+            return (value=zero);
+        }
+
+        // [Effect] Read corresponding value
+        let (contract_address) = CarbonableOffseter_carbonable_project_address_.read();
+        let (value) = IERC3525.valueOf(contract_address=contract_address, tokenId=token_id);
 
-        return (balance=balance);
+        return (value=value);
     }
 
     func total_claimed{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (
         total_claimed: felt
     ) {
-        let (users_len, users) = _read_users();
+        let (users_len) = CarbonableOffseter_users_len_.read();
 
         if (users_len == 0) {
             return (total_claimed=0);
         }
 
-        let (total_claimed) = _total_claimed_iter(index=users_len - 1, users=users);
+        let (total_claimed) = _total_claimed_iter(index=users_len - 1);
         return (total_claimed=total_claimed);
     }
 
     func total_claimable{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (
         total_claimable: felt
     ) {
-        let (users_len, users) = _read_users();
+        let (users_len) = CarbonableOffseter_users_len_.read();
 
         if (users_len == 0) {
             return (total_claimable=0);
         }
 
-        let (total_claimable) = _total_claimable_iter(index=users_len - 1, users=users);
+        let (total_claimable) = _total_claimable_iter(index=users_len - 1);
         return (total_claimable=total_claimable);
     }
 
+    func deposited_of{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        address: felt
+    ) -> (value: Uint256) {
+        let (value) = CarbonableOffseter_registered_value_.read(address);
+        return (value=value);
+    }
+
     func claimable_of{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
         address: felt
     ) -> (claimable: felt) {
         alloc_locals;
 
-        let (computed_claimable) = _claimable(user=address);
+        let (computed_claimable) = _claimable(address=address);
         let (stored_claimable) = CarbonableOffseter_claimable_.read(address);
         let (stored_claimed) = CarbonableOffseter_claimed_.read(address);
         let claimable = stored_claimable + computed_claimable - stored_claimed;
 
         // [Check] Overflow
         let (contract_address) = CarbonableOffseter_carbonable_project_address_.read();
+        let (slot) = CarbonableOffseter_carbonable_project_slot_.read();
         let (max_absorption) = ICarbonableProject.getCurrentAbsorption(
-            contract_address=contract_address
+            contract_address=contract_address, slot=slot
         );
-        let not_overflow = is_le(claimable, max_absorption);
+        let not_overflow = is_le_felt(claimable, max_absorption);
```

Comment:
I don't know how it could be, except if an attacker could manipulate block timestamp.
Do you mean we should `assert_in_range` the result of the difference?

Reply:
Done